test_runner/src/main/kotlin/ftl/run/AndroidTestRunner.kt

```diff
@@ -17,66 +18,70 @@ import kotlinx.coroutines.coroutineScope
 
 object AndroidTestRunner {
 
-    suspend fun runTests(androidArgs: AndroidArgs): MatrixMap = coroutineScope {
+    suspend fun runTests(androidArgs: AndroidArgs): Pair<MatrixMap, List<List<String>>> = coroutineScope {
         val (stopwatch, runGcsPath) = GenericTestRunner.beforeRunTests(androidArgs)
 
         // GcAndroidMatrix => GcAndroidTestMatrix
         // GcAndroidTestMatrix.execute() 3x retry => matrix id (string)
         val androidDeviceList = GcAndroidDevice.build(androidArgs.devices)
 
-        val apks = resolveApks(androidArgs, runGcsPath)
         val jobs = arrayListOf<Deferred<TestMatrix>>()
         val runCount = androidArgs.repeatTests
-        val shardCount = androidArgs.testShardChunks.size
         val shardCounter = ShardCounter()
         val history = GcToolResults.createToolResultsHistory(androidArgs)
+        val apks = resolveApks(androidArgs, runGcsPath)
+        val testShardChunks: MutableList<List<String>> = mutableListOf()
+
+        apks.forEach { apk ->
+            testShardChunks += AndroidTestShard.getTestShardChunks(androidArgs, apk.test)
+        }
 
-        println(GenericTestRunner.beforeRunMessage(androidArgs))
-        repeat(runCount) {
-            repeat(shardCount) { testShardsIndex ->
-                jobs += async {
-                    GcAndroidTestMatrix.build(
-                        appApkGcsPath = apks.first,
-                        testApkGcsPath = apks.second,
-                        runGcsPath = runGcsPath,
-                        androidDeviceList = androidDeviceList,
-                        testShardsIndex = testShardsIndex,
-                        args = androidArgs,
-                        shardCounter = shardCounter,
-                        toolResultsHistory = history
-                    ).executeWithRetry()
+        apks.forEach { apk ->
+            repeat(runCount) {
+                testShardChunks.forEach { testTargets ->
+                    jobs += async {
+                        GcAndroidTestMatrix.build(
```

Comment:
The matrix is receiving test shards that are not part of the test apk. We should create each matrix out the shards of a given apk.

```kotlin
        apks.forEach { apk ->
            val apkTestShardChunks = AndroidTestShard.getTestShardChunks(androidArgs, apk.test)
            testShardChunks.addAll(apkTestShardChunks)
            repeat(runCount) {
                apkTestShardChunks.forEach { testTargets ->
                    jobs += async {
                        GcAndroidTestMatrix.build(
                            appApkGcsPath = apk.app ?: androidArgs.appApk,
                            testApkGcsPath = apk.test,
                            runGcsPath = runGcsPath,
                            androidDeviceList = androidDeviceList,
                            testTargets = testTargets,
                            args = androidArgs,
                            shardCounter = shardCounter,
                            toolResultsHistory = history
                        ).executeWithRetry()
                    }
                }
            }
        }
```

Reply:
Fixed